docs/instrumentation/tracing-java-interop.md

```diff
@@ -0,0 +1,206 @@
+```
+
+1) `Context.wrap(ctx)` - creates otel4s context from the `JContext`  
+2) `Local[F, Context].scope` - sets the given context as an active environment for the effect `fa`
+
+_____
+
+Let's say you use [Pekko HTTP][pekko-http] and want to materialize an `IO` using the current tracing context: 
+```scala mdoc:silent:reset
+import cats.effect.{Async, IO}
+import cats.effect.std.Random
+import cats.effect.syntax.temporal._
+import cats.effect.unsafe.implicits.global
+import cats.mtl.Local
+import cats.syntax.all._
+import org.apache.pekko.http.scaladsl.model.StatusCodes.OK
+import org.apache.pekko.http.scaladsl.server.Directives._
+import org.apache.pekko.http.scaladsl.server.Route
+import org.typelevel.otel4s.Attribute
+import org.typelevel.otel4s.trace.Tracer
+import org.typelevel.otel4s.java.context.Context
+import io.opentelemetry.instrumentation.annotations.WithSpan
+import io.opentelemetry.context.{Context => JContext}
+import scala.concurrent.duration._
+
+def route(implicit T: Tracer[IO], L: Local[IO, Context]): Route = 
+  path("gen-random-name") {
+    get {
+      complete {
+        OK -> generateRandomName(length = 10)
+      }
+    }
+  }
+
+@WithSpan("generate-random-name")
+def generateRandomName(length: Int)(implicit T: Tracer[IO], L: Local[IO, Context]): String =
+  withJContext(JContext.current())(generate[IO](length)).unsafeRunSync()
+
+def generate[F[_]: Async: Tracer](length: Int): F[String] =
+  Tracer[F].span("generate", Attribute("length", length.toLong)).surround {
+    for {
+      random <- Random.scalaUtilRandom[F]
+      delay  <- random.betweenInt(100, 2000)
+      chars  <- random.nextAlphaNumeric.replicateA(length).delayBy(delay.millis)
+    } yield chars.mkString
+  }
+
+def withJContext[F[_], A](ctx: JContext)(fa: F[A])(implicit L: Local[F, Context]): F[A] =
+  Local[F, Context].scope(fa)(Context.wrap(ctx))
+```
+
+When you invoke the `gen-random-name` endpoint, the spans will be structured in the following way:
+```
+> GET { http.method = GET, http.target = /gen-random-name, ... }
+  > generate-random-name 
+    > generate { length = 10 } 
+```
+
+## How to use otel4s context with OpenTelemetry Java
+
+To interoperate with Java libraries that rely on the OpenTelemetry Java context, you need to activate the context manually.
+The following utility method allows you to extract the current otel4s context and set it into the ThreadLocal variable:
+
+```scala mdoc:silent:reset
+import cats.effect.Sync
+import cats.mtl.Local
+import cats.syntax.flatMap._
+import org.typelevel.otel4s.java.context.Context
+import io.opentelemetry.context.{Context => JContext}
+import scala.util.Using
+
+def useJContext[F[_]: Sync, A](use: JContext => A)(implicit L: Local[F, Context]): F[A] = 
+  Local[F, Context].ask.flatMap { ctx => // <1>
+    Sync[F].defer {
+      Sync[F].fromTry {
+        val jContext: JContext = ctx.underlying // <2>
+        Using(jContext.makeCurrent())(_ => use(jContext)) // <3>
+      }
+    }
```

Comment:
`defer`/`fromTry` is an unusual pattern (`defer` is actively encouraged to be ignored on Discord). It also doesn't translate well to e.g. `blocking`. Instead of the `Using` utility I wonder if a simple `try`/`finally` inside a `delay`/`blocking` would be better.

Reply:
Valid point. `try/finally` should fit better.